### PR TITLE
feat(pacquet): report extraneous packages in pnpm list

### PR DESCRIPTION
**Pull request by zkochan**
## Summary

`pnpm list --json` and `pnpm list --parseable` now report **extraneous** packages in pacquet — packages present in an importer's `node_modules` but absent from the lockfile — under an `unsavedDependencies` object, matching the TypeScript CLI.

pacquet builds the list tree entirely from the lockfile, so extraneous packages (an npm-installed package later removed from `package.json`, or an unrecorded local link) were never surfaced. This closes that parity gap. The gap was spotted while reviewing pnpm/pnpm#12866 (the TypeScript-side EMFILE fix for reading these same unsaved dependencies).

Behavior matches the TypeScript CLI exactly:

- The **default tree view still omits** extraneous packages — the TypeScript CLI hardcodes `showExtraneous: false`, and only its `--json` and `--parseable` renderers surface them. pacquet does the same: the tree renderer is untouched; only the JSON and parseable renderers gain the new group.
- A symlinked entry (a workspace / `link:` dependency the manifest no longer records) reports `version: "link:<relative-to-project>"` and its resolved path; a regular directory reports its `package.json` version (or the literal `undefined` when missing/unreadable) and its on-disk path — the shape `buildDependenciesTree` produces.
- The `node_modules` scan follows the same rules as the TypeScript `readModulesDir`: skip files and dot-prefixed entries (`.pnpm`, `.bin`), descend one level into `@scope` directories.
- Detection is skipped when searching by package name and for `--depth -1` (project-only output) — the same two gates the TypeScript CLI applies before it scans `node_modules`.
- The "saved" set spans all three importer dependency groups regardless of `--prod`/`--dev`/`--optional`, matching `getAllDirectDependencies`.

No new dependencies: reuses `pacquet_fs::lexical_normalize` (matching `resolve-link-target`'s lexical resolution) and `pathdiff`, both already in the crate.

Known limitation: `std::fs::read_link` does not resolve Windows junctions (which the linker uses there), so an unsaved *workspace-link* package on Windows would report its `package.json` version rather than `link:<target>` — a rare, non-breaking edge case; the symlink unit test is unix-gated.

## Squash Commit Body

```text
feat(pacquet): report extraneous packages in pnpm list

`pnpm list --json` and `pnpm list --parseable` now report packages that
are present in an importer's `node_modules` but absent from the lockfile
(npm's "extraneous" dependencies) under `unsavedDependencies`, matching
the TypeScript CLI.

pacquet built the list tree entirely from the lockfile, so extraneous
packages were never surfaced. The tree is still lockfile-only; the scan
of `node_modules` runs only for the `--json` and `--parseable` renderers,
mirroring the TypeScript CLI, which hardcodes `showExtraneous: false` for
the default tree view. Extraneous detection is skipped when searching by
package name and for `--depth -1` (project-only output) — the same two
gates the TypeScript CLI applies before scanning `node_modules`.
```

## Checklist

- [x] The change is implemented in both the TypeScript CLI and the Rust
  `pnpm/` port, or the description notes what still needs porting. *(The TypeScript CLI already ships this behavior; this PR brings the Rust `pnpm/` port to parity.)*
- [x] Added a changeset (`pnpm changeset`) if this PR changes any published
  package. Keep it short and written for pnpm users — it becomes a release note.
- [x] Added or updated tests.

---
Written by an agent (Claude Code, claude-opus-4-8).

---
<a href="https://app.blacksmith.sh/pnpm/codesmith/pnpm/pr/12967"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1786534649&installation_id=145934176&pr_number=12967&repository=pnpm%2Fpnpm&return_to=https%3A%2F%2Fgithub.com%2Fpnpm%2Fpnpm%2Fpull%2F12967&signature=4d44c8221ee037670db17798ee346cc885650f1362426ec79cf9a82aeec91aec"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

* **New Features**
  * `pnpm list --json` now reports packages installed in `node_modules` but missing from the lockfile under `unsavedDependencies`.
  * `pnpm list --parseable` now includes these unsaved packages in its output.
* **Bug Fixes**
  * Package listings now more accurately distinguish lockfile dependencies from additional installed packages.

### CODE REVIEW

File: pnpm/crates/cli/src/cli_args/list.rs [hunk 1]
```diff
@@ -588,6 +607,140 @@ pub(crate) fn read_root_manifest(
...
+        }
+    };
+    Some(lexical_normalize(&joined))
+}
+
+fn read_package_version(pkg_dir: &Path) -> Option<String> {
+    let bytes = std::fs::read(pkg_dir.join("package.json")).ok()?;
+    let manifest: Value = serde_json::from_slice(&bytes).ok()?;
+    manifest.get("version").and_then(Value::as_str).map(str::to_string)
+}
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/cli/src/cli_args/list.rs]: Good point — per the repo convention ("Tests are documentation"), I trimmed the enumerated scenarios from the doc comments on `build_unsaved_node` and `read_modules_dir_names`, since the new unit tests already assert them. Kept `resolve_link_target`'s doc to the non-obvious *why* (lexical, existence-free resolution matching `resolveLinkTarget`). Fixed in 730ae8a695.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pnpm/crates/cli/src/cli_args/list.rs [hunk 2]
```diff
@@ -219,6 +221,16 @@ impl ListArgs {
...
+        // Extraneous packages are suppressed when searching by name (a
+        // package-name argument narrows to lockfile matches) and when
+        // `--depth -1` asks for project-only output — the same two gates
+        // the TypeScript CLI applies before it scans `node_modules`.
+        let unsaved = if self.packages.is_empty() && self.depth != RecursionLimit::ProjectsOnly {
+            read_unsaved_dependencies(importer, dir, config)
+        } else {
+            Vec::new()
+        };
+
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/cli/src/cli_args/list.rs]: Applied — good catch. The scan now runs only when `--json` or `--parseable` will surface the result, so plain `pnpm list` / `pnpm -r list` (the common path) no longer pays for the `node_modules` scan. Fixed in 730ae8a695.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pnpm/crates/cli/src/cli_args/list.rs [hunk 3]
```diff
@@ -588,6 +607,140 @@ pub(crate) fn read_root_manifest(
...
+    };
+    Some(lexical_normalize(&joined))
+}
+
+fn read_package_version(pkg_dir: &Path) -> Option<String> {
+    let bytes = std::fs::read(pkg_dir.join("package.json")).ok()?;
+    let manifest: Value = serde_json::from_slice(&bytes).ok()?;
+    manifest.get("version").and_then(Value::as_str).map(str::to_string)
+}
+
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/cli/src/cli_args/list.rs:757]: Applied. `read_modules_dir_names`/`collect_module_names` now return `io::Result`: a missing directory (`ENOENT`) yields no entries, while any other read failure is surfaced — matching `readModulesDir`, which only swallows `ENOENT`. Fixed in 730ae8a695.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: .changeset/pacquet-list-unsaved-deps.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+"pacquet": patch
+---
```
[the remark this answers is not in the document]
Comment (zkochan) [reply .changeset/pacquet-list-unsaved-deps.md:3]: This one is a false positive for this PR. Per the repo's changeset policy (AGENTS.md, "Changesets for the Rust products"): in a changeset, `"pnpm"` always means the **TypeScript** CLI package, and *"Do not add `"pnpm"` to a Rust-only changeset."* This change is pacquet-only — the TypeScript CLI already ships this behavior and no TypeScript package is modified here — so the changeset correctly targets only `"pacquet"`. Adding `"pnpm"` would incorrectly bump the TypeScript CLI for a change it does not contain. Leaving the changeset as-is.

---
Written by an agent (Claude Code, claude-opus-4-8).
